---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# cortar

<!-- badges: start -->
<!-- badges: end -->

The goal of cortar is to extract reads at the exon-intron junction from RNA-seq
and report the proportion of splicing events across each intron. By comparing
these values to controls, deviations from normal splicing can be characterised
in test samples.

## Installation

You can install the development version of cortar from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("kidsneuro-lab/RNA_splice_tool")
```

## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
#library(cortar)
## basic example code
```

To use cortar, a samplefile needs to be created for each run. This file
contains six columns and a row for each sample (as shown below):
* sampleID: A unique identifier for the sample
* familyID: A unique identifier for related samples that should not be compared
to one another
* sampletype: Whether analysis and report is desired for this sample ("test")
* genes: Gene symbol for gene under investigation (can be blank for
panel/research mode)
* transcript: Transcript under investigation (blank transcript will default to
the canonical RefSeq transcript)
* bamfile: Absolute or relative file path to the processed bamfile of the sample

```{r cars}
print(example_cortar_samplefile)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/v1/examples>.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
